### ISSUE TITLE
With dbt templater, `sqlfluff fix <file>.sql` will overwrite lines if rule LO34 is triggered

### ISSUE THREAD

**Issue by 808sAndBR** [sqlfluff/sqlfluff#2117]
## Expected Behaviour

`sqlfluff fix <file>.sql` should fix any formatting issues in place. 

## Observed Behaviour

`sqlfluff fix <file>.sql` will drop line and overwrite them with the code that is present in the macro, if rule LO34 is triggered and a macro exists in the code. This can occur when the macro contains parenthesis, if there is a `*` on a line other than the first one, or assumably other ways of triggering this rule that I am unaware of.

As a work around rule LO34 can be skipped with:

```
sqlfluff fix <file or dir> --exclude-rules L034
``` 

I believe this to be the root cause of the issues in:

* https://github.com/sqlfluff/sqlfluff/issues/1952
* https://github.com/sqlfluff/sqlfluff/issues/2056
* https://github.com/sqlfluff/sqlfluff/issues/1953

I also suspect it to be the cause of this one but don't have quite enough context for me to 100% convince myself:

* https://github.com/sqlfluff/sqlfluff/issues/2065

### Works as expected:

model.sql:
```
SELECT
   {{ normal() }}
  , line_two As line_two
FROM foo
```

macro.sql:
```
{% macro normal() %}
     col AS col
{% endmacro %}
```

`sqlfluff fix model.sql` makes model.sql:
```
SELECT
   {{ normal() }}
  , line_two AS line_two
FROM foo
```

### parenthesis wrapped (unexpected)

model.sql:
```
SELECT
   {{ par_wrap() }}
  , line_two As line_two
FROM foo
```

macro.sql:
```
{% macro par_wrap() %}
    ( col ) AS col
{% endmacro %}
```

`sqlfluff fix model.sql` makes model.sql:
```
SELECT
   {{ par_wrap() }}
  , ( col ) AS col
FROM foo
```

### in list (unexpected)

model.sql:
```
SELECT
   {{ in_list() }}
  , line_two As line_two
FROM foo
```

macro.sql:
```
{% macro in_list() %}
     col in ('var1', 'var2')
{% endmacro %}
```

`sqlfluff fix model.sql` makes model.sql:
```
SELECT
   {{ in_list() }}
  , col IN ('var1', 'var2')
FROM foo
```

A select * triggering Rule LO34

```
SELECT
   {{ normal() }}
  , line_two As line_two
  , *
FROM foo
```

macro.sql:
```
{% macro normal() %}
     col AS col
{% endmacro %}
```

`sqlfluff fix model.sql` makes model.sql:
```
SELECT
   {{ normal() }}
  , col AS col
  , second_line AS second_line
FROM foo
```

Similarly able to have this occur if the macro uses a sql functions that has parenthesis.

## Dialect

snowflake

## Version

sqlfluff: 0.9.0
sqlfluff-templater-dbt: 0.9.0
dbt:  0.21.1

I also tested this a few different versions of everything and it happened consistently.

## Configuration

```
[sqlfluff]
templater = dbt
dialect = snowflake

[sqlfluff:rules]
tab_space_size = 2
comma_style = leading

[sqlfluff:rules:L010]
capitalisation_policy = upper

[sqlfluff:templater:dbt]
profile = dbt_snowflake
```

(I believe the `templater = dbt` is the only part relevant for this bug but including all of it as is, just in case)

**Comment by 808sAndBR** [comment]
Also if helpful, here are the debug logs I get out from passing the build command -vvvv

```
DEBUG      Original Tree: 'SELECT\n   \n     lower(col)\n\n  , line_two As line_two\nFROM foo\n'
DEBUG      Fixed Tree: 'SELECT\n   \n  line_two AS line_two\n\n  , lower(col)\nFROM foo\n'
DEBUG      ### Templated File.
DEBUG          File slice: 0 TemplatedFileSlice(slice_type='literal', source_slice=slice(0, 10, None), templated_slice=slice(0, 10, None)) [invariant]
DEBUG          File slice: 1 TemplatedFileSlice(slice_type='templated', source_slice=slice(10, 25, None), templated_slice=slice(10, 27, None))
DEBUG          			templated: '\n     lower(col)\n'	source: '{{ in_list() }}'
DEBUG          File slice: 2 TemplatedFileSlice(slice_type='literal', source_slice=slice(25, 60, None), templated_slice=slice(27, 62, None)) [invariant]
DEBUG          File slice: 3 TemplatedFileSlice(slice_type='literal', source_slice=slice(59, 60, None), templated_slice=slice(61, 62, None)) [invariant]
DEBUG      Source-only slices: []
DEBUG      ### Beginning Patch Iteration.
DEBUG      FileSegment at slice(0, 62, None): Original: ['SELECT\n   \n     lower(col)\n\n  , line_two As line_two\nFROM foo\n'] Fixed: ['SELECT\n   \n  line_two AS line_two\n\n  , lower(col)\nFROM foo\n']
DEBUG      StatementSegment at slice(0, 61, None): Original: ['SELECT\n   \n     lower(col)\n\n  , line_two As line_two\nFROM foo'] Fixed: ['SELECT\n   \n  line_two AS line_two\n\n  , lower(col)\nFROM foo']
DEBUG      SelectStatementSegment at slice(0, 61, None): Original: ['SELECT\n   \n     lower(col)\n\n  , line_two As line_two\nFROM foo'] Fixed: ['SELECT\n   \n  line_two AS line_two\n\n  , lower(col)\nFROM foo']
DEBUG      SelectClauseSegment at slice(0, 52, None): Original: ['SELECT\n   \n     lower(col)\n\n  , line_two As line_two'] Fixed: ['SELECT\n   \n  line_two AS line_two\n\n  , lower(col)']
DEBUG      Appending insertion buffer. '  ' @idx: 11
DEBUG      Appending insertion buffer. '  line_two AS line_two' @idx: 11
DEBUG        0 Yielded patch: FixPatch(templated_slice=slice(11, 26, None), fixed_raw='  line_two AS line_two', patch_category='mid_point')
DEBUG              Templated Hint: ...'ELECT\n   \n' <> '\n\n  , line'...
INFO             - Skipping patch over templated section: EnrichedFixPatch(source_slice=slice(10, 25, None), templated_slice=slice(11, 26, None), fixed_raw='  line_two AS line_two', patch_category='mid_point', templated_str='     lower(col)', source_str='{{ in_list() }}')
DEBUG      Appending insertion buffer. 'lower(col)' @idx: 32
DEBUG        1 Yielded patch: FixPatch(templated_slice=slice(32, 52, None), fixed_raw='lower(col)', patch_category='end_point')
DEBUG              Templated Hint: ...'col)\n\n  , ' <> '\nFROM foo\n'...
INFO             * Keeping patch on literal-only section: EnrichedFixPatch(source_slice=slice(30, 50, None), templated_slice=slice(32, 52, None), fixed_raw='lower(col)', patch_category='end_point', templated_str='line_two As line_two', source_str='line_two As line_two')
DEBUG      Final slice buffer: [slice(0, 30, None), slice(30, 50, None), slice(50, 60, None)]
DEBUG      Appending Raw:                    slice(0, 30, None)     'SELECT\n   {{ in_list() }}\n  , '
DEBUG      Appending end_point Patch:        slice(30, 50, None)    'line_two As line_two' > 'lower(col)'
DEBUG      Appending Raw:                    slice(50, 60, None)     '\nFROM foo\n'
```

In this part it almost seems like its mixing up the templated section and the literal-only section maybe? Still super new to digging into this 🤔 

```
INFO             - Skipping patch over templated section: EnrichedFixPatch(source_slice=slice(10, 25, None), templated_slice=slice(11, 26, None), fixed_raw='  line_two AS line_two', patch_category='mid_point', templated_str='     lower(col)', source_str='{{ in_list() }}')
```

**Comment by aa0910** [comment]
@808sAndBR Hi and thank you for flagging the root cause. My team identified the issues #1952 and #1953 and we were wondering if there is a workaround this in dbt / sqlfluff that you found yet? Really appreciate it. This is the only thing that is holding us up in deploying sqlfluff in our team.

**Comment by 808sAndBR** [comment]
@aa0910 I havent figured out a work around or fix yet, but I also have not had a chance to dig in much. We are similarly pretty much ready to use this once we get it sorted out and i'm super excited for this tool so potentially trying to block off some time to dig in more soon if no one with more domain knowledge of the package chimes in with potential causes or work arounds :) 

If you figure anything out though I am all ears!!

**Comment by aa0910** [comment]
@808sAndBR thank you! Yes we are in the same situation. Will get back if we find something and would love to know if you find a solution/ workaround as well. Appreciate it.

**Comment by 808sAndBR** [comment]
I have dug into this a good bit today and have narrowed it down to being an issue with how the reordering works from Rule LO34. (I have added an example in the issue description where this can be triggered without having parenthesis in the macro)

As far as I can tell, lines with macros have something where they override changes that would be applied to their compiled versions, however LO34's reordering seems to reorder the compiled version and pass compiled code on. Lines that include a macro will then over write the compiled code's lines staying in their original position (I _think_) while the code compiled from the macro will now be exposed in the file as raw code. 

This part of the codebase seems fairly complex so I am having a hard time figuring out if the fix is a complicated rework of the logic or if its something more like a raw vs compiled tag didn't get applied somewhere 🤔. Would love thoughts from someone with more experience in this. 

In the meantime running `sqlfluff fix <file or dir> --exclude-rules L034` seems to let me build with expected changes as a temporary work around. (@aa0910 please also try this workaround for your use case and see if it helps.)

**Comment by aa0910** [comment]
thank you @808sAndBR which version of dbt are you using? I am using the latest 1.0.0 and sqlfluff does not seem to run with it.

**Comment by aa0910** [comment]
@808sAndBR thanks a lot for looking into this. This workaround works well for me as well.

**Comment by barrywhart** [comment]
SQLFluff tries to avoid making changes inside templated code. The existing logic works well for rules that fix code "in place". It can make mistakes with fixes that move code around.

IIUC, the policy is based on whether the "anchor" segment of a fix is inside a templated region. We may need to look at expanding this, e.g. to look at all the segments touched by a fix, not just the anchors. This issue may be easier to fix once #2126 has been merged -- it provides easier ways to work with collections of segments and raw/templated slices.

**Comment by barrywhart** [comment]
Note this is a core SQLFluff issue, not specific to the dbt templater. I reproduced the first issue today using the jinja templater:
```
SELECT
   {{ par_wrap() }}
  , line_two AS line_two
FROM foo
```

`.sqlfluff`:
```
[sqlfluff:templater:jinja:macros]
par_wrap = {% macro par_wrap() %}
    ( col ) AS col
    {% endmacro %}
```

**Comment by barrywhart** [comment]
Heh, I think there are several bugs here. I made some improvements to the "does this touch templated code" check, but there are still issues. I think L034 may be "seeing" the newline in the macro definition and doing something weird. Very cool bug! I'll do my best to find and solve all the issues here, but it may take a few days. 😅

**Comment by barrywhart** [comment]
@808sAndBR, @aa0910: This is fixed now. We're planning a new release in the next couple of days which will include the fix.

**Comment by 808sAndBR** [comment]
@barrywhart This is awesome!! Thanks so much for digging into this and getting it sorted out ❤️
